torch/csrc/jit/codegen/cuda/ir_nodes.cpp

```diff
@@ -1031,6 +1030,279 @@ std::pair<TensorDomain*, TensorDomain*> TensorDomain::rFactor(
+
+  // utility to record new concretize opportunity
+  void concretizeTo(IterDomain* id, IterDomain* To) {
+    TORCH_INTERNAL_ASSERT(id->isBroadcast() && !To->isBroadcast());
+    BcastDomainMap_[id] = concretized(To);
+  }
+
+  void handle(UnaryOp* uop) override {
+    concretizePwOp(uop->asExpr());
+  }
+
+  void handle(BinaryOp* bop) override {
+    concretizePwOp(bop->asExpr());
+  }
+
+  void handle(TernaryOp* top) override {
+    concretizePwOp(top->asExpr());
+  };
+}; // class ConcretizeDomain
+
+void ConcretizeDomain::concretizePwOp(Expr* e) {
+  TensorView* TVo = *ir_utils::filterByType<TensorView>(e->outputs()).begin();
+
+  // won't use this function on a reduction op, so root domain only
+  std::vector<IterDomain*> Io = TVo->getRootDomain();
+
+  for (auto* i : ir_utils::filterByType<TensorView>(e->inputs())) {
+    std::vector<IterDomain*> Ii = i->getRootDomain();
+    TORCH_INTERNAL_ASSERT(Ii.size() == Io.size());
+
+    for (size_t it = 0; it < Ii.size(); it++) {
+      if (!canConcretize(Io[it]))
+        continue;
+
+      if (!canConcretize(Ii[it]))
+        concretizeTo(Ii[it], concretized(Io[it]));
+    }
+  }
+} // ConcretizeDomain::pointWiseOpMap
+
+/** \brief Models equality provable by the graph
+ *  \details This traversal processes root domains only,
+ *           equalities , e.g. :
+ *           T2 [i0,i1] = T1[i2,i3] + T0[i4,i5]
+ *           will prove that i2 and i4 are equal in the sense that
+ *             i2.start = i4.start, i2.extent = i4.extent
+ *
+ *           Depends on ConcretizeDomain, and equalities involving
+ *           broadcast domains are defined based on the concretized version
+ */
+
+class ProveValEqual : public IterVisitor {
+ public:
+  ProveValEqual(Fusion* fusion) : CD_(fusion) {
+    traverse(fusion, false);
+  }
+
+  /** \brief checks if two scalars are equal
+   *  \details first checks if ScalarCheck has them equal,
+   *           next try to prove them equal from
+   *           the graph_traversal result
+   */
+  bool areEqual(Val* a, Val* b) {
+    if (ScalarCheck::sameAs(a, b))
+      return true;
+    if (EqSet_.areEqual(a, b))
+      return true;
+    return false;
+  }
+
+  /** \brief checks if two iterdomains are equal
+   *  \details equality defined as equal start and equal extent
+   *           true means a and b are equal
+   *           false only means that they cannot be proven equal based
+   *           on scalar check and graph traversal
+   */
+  bool areEqual(IterDomain* a, IterDomain* b) {
+    if (a->sameAs(b))
+      return true;
+
+    // abort on un-concretized domains, this can appear once we
+    // allow broadcast on fusion output
+    if (!CD_.canConcretize(a) || !CD_.canConcretize(b))
+      return false;
+
+    auto ac = CD_.concretized(a);
+    auto bc = CD_.concretized(b);
+    return (
+        areEqual(ac->start(), bc->start()) &&
+        areEqual(ac->rawExtent(), bc->rawExtent()));
+  }
+
+ protected:
+  ConcretizeDomain CD_;
```

Comment:
I think our naming convention prefers `cd_` and `eqSet_`.

Reply:
Thanks for pointing this out. Just updated.